### ISSUE TITLE
More "safename"

### ISSUE THREAD

**Issue by hhsprings** [httplib2/httplib2#27]
This issue is maybe of Windows issue. Newline in filename is likely the problem.

``` python
>>> import httplib2
>>> http = httplib2.Http(".cache")
>>> headers, contents = http.request("https://skyvector.com/airport/83Q", "GET")
>>> headers, contents = http.request("https://skyvector.com/airport/83Q", "GET")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "c:\Python27\lib\site-packages\httplib2-0.9-py2.7.egg\httplib2\__init__.py", line 1532, in request
    headers=headers, redirections=redirections - 1)
  File "c:\Python27\lib\site-packages\httplib2-0.9-py2.7.egg\httplib2\__init__.py", line 1593, in request
    (response, content) = self._request(conn, authority, uri, request_uri, method, body, headers, redirections, cachekey)
  File "c:\Python27\lib\site-packages\httplib2-0.9-py2.7.egg\httplib2\__init__.py", line 1396, in _request
    _updateCache(headers, response, content, self.cache, cachekey)
  File "c:\Python27\lib\site-packages\httplib2-0.9-py2.7.egg\httplib2\__init__.py", line 448, in _updateCache
    cache.set(cachekey, text)
  File "c:\Python27\lib\site-packages\httplib2-0.9-py2.7.egg\httplib2\__init__.py", line 715, in set
    f = file(cacheFullPath, "wb")
IOError: [Errno 22] invalid mode ('wb') or filename: '.cache\\skyvector.com,airport,83Q,Port-of-Poulsbo-Marina-Moorage-Seaplane\r\n Base,f2669977249d54dcf311752972d65d2f'
>>>
```

`safename` in `__init__.py` should replace newline, I think.

**Comment by galvez** [comment]
@hhsprings Guessing you likely already monkeypatched this, but I'll double check the behavior on the latest `master` and push a fix if still needed.

**Comment by hhsprings** [comment]
> Guessing you likely already monkeypatched this

My solution was minimal and dirty...:
```diff
--- c:/Python27/Lib/site-packages/httplib2-0.9-py2.7.egg/httplib2/__init__.py~	2015-02-10 01:58:24 +0900
+++ c:/Python27/Lib/site-packages/httplib2-0.9-py2.7.egg/httplib2/__init__.py	2016-09-12 13:01:09 +0900
@@ -252,7 +252,7 @@
         filename=filename.encode('utf-8')
     filemd5 = _md5(filename).hexdigest()
     filename = re_url_scheme.sub("", filename)
-    filename = re_slash.sub(",", filename)
+    filename = re_slash.sub(",", filename).replace("\r\n", "\n").replace("\n", " ")
 
     # limit length of filename
     if len(filename)>200:
```

**Comment by axil** [comment]
@hhsprings Ampersand is also a problem (I ran into this when debugging an issue I ran into in openid module). I'd suggest using `urllib.parse.quote` to be 100% sure.

**Comment by axil** [comment]
Now it is maximum filename length, which is 259.   
```
C:\Users\    <username>    \AppData\Local\Temp\  <safe_filename>  ,   <md5>
  9 chars + max 104 chars  +     20 chars      +       x       +  1  +  32  = max 259 chars
```
So max safe filename x = 93 chars, let it be 90 to make a round sum:

```diff
diff --git a/python3/httplib2/__init__.py b/python3/httplib2/__init__.py
index 915fd7e..71f6f7c 100644
--- a/python3/httplib2/__init__.py
+++ b/python3/httplib2/__init__.py
@@ -181,14 +181,15 @@ def safename(filename):
         pass
     if isinstance(filename,str):
         filename=filename.encode('utf-8')
-    filemd5 = _md5(filename).hexdigest().encode('utf-8')
+    filemd5 = _md5(filename).hexdigest()
     filename = re_url_scheme.sub(b"", filename)
-    filename = re_slash.sub(b",", filename)
+    #filename = re_slash.sub(b",", filename)
+    filename = urllib.parse.quote(filename, safe='')

     # limit length of filename
-    if len(filename)>200:
-        filename=filename[:200]
-    return b",".join((filename, filemd5)).decode('utf-8')
+    if len(filename) > 90:
+        filename = filename[:90]
+    return ",".join((filename, filemd5))

 NORMALIZE_SPACE = re.compile(r'(?:\r\n)?[ \t]+')
 def _normalize_headers(headers):
```

**Comment by hhsprings** [comment]
To shorten the filename, we can use base64 module, i think:
```python
    hasher = hashlib.new("md5")
    hasher.update(fn.encode("utf-8"))
    hashname = re.sub(r"=+$", "", base64.urlsafe_b64encode(
        hasher.digest()).decode())
    #hashname = re.sub(r"=+$", "", base64.b32encode(
    #    hasher.digest()).decode())
```

**Comment by temoto** [comment]
@hhsprings sorry friend, this is wrong in few senses.

We came to agreement with @axil in his patch. Maybe we wait for each other to continue forward. Maybe he's just busy to make final change before merge.

**Comment by temoto** [comment]
Fix is merged in 56b1fa36be8c3f251602b6be8486c7b891085cce
